Remove legacy browsers Frame Breaking js code usage and deprecate it

andrepereiradasilva opened the pull request:
### Summary of Changes

Remove Legacy Browser (IE 7 and older)  Frame Breaking Script.

Since Joomla 3.x only support IE8+ and modern browsers this code have no effect anymore because all this browsers support the `X-Frame-Options` header that already does this.

See https://www.owasp.org/index.php/Clickjacking_Defense_Cheat_Sheet

This code was only used in admin com_login page:
- isis: https://github.com/joomla/joomla-cms/blob/staging/administrator/components/com_login/views/login/tmpl/default.php#L12
- hathor: https://github.com/joomla/joomla-cms/blob/staging/administrator/templates/hathor/login.php#L23

So this PR deprecates the behaviour.noframes and in replace adds and HTTP Header in com_login admin view html file.
### Testing Instructions
1. Apply patch
2. Admin login in isis and hathor work without issues and have the `X-Frame-Options` HTTP header
3. Code review
### Documentation Changes Required

None.

andrepereiradasilva committed:
remove legacy frame breaking js code

andrepereiradasilva commented:
need some help with the unit tests.
JHtmlBehaviorTest::testNoFrames is failing and don't really know how to rewrite it.

bertmert commented on libraries/cms/html/behavior.php:
```diff
@@ -800,28 +800,6 @@ public static function noframes()
-			}
-
-			// Firefox fix
-			jQuery("input[autofocus]").focus();
```
Don't we need it anymore? Maybe someone can replace it by vanilla JS?

andrepereiradasilva replied:
Whats is that fix supposed to do?

mbabker commented:
JHtmlBehaviorTest::testNoFrames is validating that JHtmlBehavior::core gets
called.  You're removing the call to it so remove that from the array
that's being validated.

For the other failures, I'm leaning toward that being a result of the JHtml
tests affecting global state and a value not being reset correctly.

On Friday, August 19, 2016, andrepereiradasilva <notifications@github.com>
wrote:

> need some help with the unit tests.
> JHtmlBehaviorTest::testNoFrames is failing and don't really know how to
> rewrite it.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/joomla/joomla-cms/pull/11679#issuecomment-241168026>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAWfobN_3pgi0skKvNghSpHltR1xYD3jks5qhlDngaJpZM4Jo9Ir>
> .
>

dgt41 commented on libraries/cms/html/behavior.php:
```diff
@@ -800,28 +800,6 @@ public static function noframes()
 			return;
 		}
 
```
@andrepereiradasilva Can you also put depreciation tag on this function?

dgt41 commented:
@andrepereiradasilva B/C policy states that this can not be done in 3.x. What could be done is to remove the calls to this function and deprecate the function so it will be removed in j4

andrepereiradasilva commented:
@dgt41 
> B/C policy states that this can not be done in 3.x.

Why do you think this is not B/C? I'm not removing the method (the setHeader is still there), just some obslete js code inside the method.

dgt41 commented:
Ah, ok didn't catch that, still I think deprecating this and adding 
```php
JFactory::getApplication()->setHeader('X-Frame-Options', 'SAMEORIGIN');
```
to the two views is safer, but I don't mind either way

andrepereiradasilva commented:
ok have no issue with that and also think it's better.
will check.

andrepereiradasilva committed:
add to the view file

andrepereiradasilva committed:
remove behavior.noframes from template

andrepereiradasilva committed:
Remove behavior.noframes from hathor template

andrepereiradasilva committed:
restore noframes behaviour and deprecate it on 4.0

andrepereiradasilva committed:
remove extra space

andrepereiradasilva committed:
missing tab

andrepereiradasilva commented:
ok @dgt41:
- no change in method. just added deprected messages
- remove method usage where it was used in the core
- added the X-Frame-Option header to admin com_login view html file.

please test.

dgt41 commented:
I have tested this item :white_check_mark: successfully on b417a547486432f31bb3682491d9323756b60991<br /><br />Tested on OSX with: safari, chrome, ff, opera (latest, development, tech review)<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/11679">issues.joomla.org/joomla-cms/11679</a>.</sub>

brianteeman commented:
I have tested this item :white_check_mark: successfully on b417a547486432f31bb3682491d9323756b60991<br /><br />I didnt even know we had this protection - cool;)

<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/11679">issues.joomla.org/joomla-cms/11679</a>.</sub>

dgt41 commented:
RTC

bertmert commented:
You ignored the FF fix that is missing now for login form.
`// Firefox fix
jQuery("input[autofocus]").focus();`

https://github.com/joomla/joomla-cms/pull/8723

wilsonge commented:
Merged with https://github.com/joomla/joomla-cms/commit/92c08ef4e40574c9c51568da793e38ace461bb23
